src/storage/src/hummock/store/memtable.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+use std::ops::RangeBounds;
+
+use bytes::Bytes;
+
+macro_rules! memtable_impl_method_body {
+    ($memtable:expr, $method_name:ident, $($args:expr),+) => {
+        {
+            match $memtable {
+                Memtable::BTree(inner) => {
+                    inner.$method_name($($args),+)
+                },
+            }
+        }
+    };
+}
+
+// TODO: may want to change the return type to avoid cloning Bytes.
+pub type MemtableIter = impl Iterator<Item = (Bytes, Bytes)>;
+
+pub enum Memtable {
+    BTree(BTreeMapMemtable),
+}
+
+#[allow(unused)]
+impl Memtable {
+    /// Inserts a key-value entry associated with a given `epoch` into memtable.
+    fn insert(&mut self, key: Bytes, val: Bytes, epoch: u64) {
+        memtable_impl_method_body!(self, insert, key, val, epoch)
+    }
+
+    /// Deletes a key-value entry from memtable. Only the key-value entry with epoch smaller
+    /// than the given `epoch` will be deleted.
+    fn delete(&mut self, key: Bytes, epoch: u64) {
```

Comment:
Actually, just to be clear - shouldn't we only delete entries with exact same epoch, as we might still want a consistent snapshot of previous epochs to facilitate recovery of that epoch?

Reply:
> shouldn't we only delete entries with exact same epoch

No. Consider the following operations:
```
epoch1: insert k1, v1
epoch2: delete k1
```

It is possible that there is no entry written with the exact same epoch of the delete operation. In the above example, k1 should be invisible after epoch2. Btw, in order to achieve MVCC on memtable, deletion is implemented as writing a tombstone key so reading with older epoch can still see consistent result.

Reply:
Indeed, I was confusing the statement as a physical delete rather than as logical. Thanks for clarifying.